nutils/cli.py

```diff
@@ -1,397 +1,361 @@
 """
-The cli (command line interface) module provides the `cli.run` function that
+he cli (command line interface) module provides the `cli.run` function that
 can be used set up properties, initiate an output environment, and execute a
 python function based arguments specified on the command line.
 """
 
-from . import long_version, warnings, util
-import sys
+import bottombar
+import contextlib
+import functools
 import inspect
 import os
-import time
-import signal
-import subprocess
-import contextlib
-import traceback
 import pathlib
-import html
-import functools
-import pdb
 import stringly
-import textwrap
-import typing
+import sys
 import treelog
-import collections
 
-try:
-    Level = treelog.proto.Level
-except AttributeError:  # treelog version < 1.0b6
-    Level = collections.namedtuple('Level', ['debug', 'info', 'user', 'warning', 'error'])(0, 1, 2, 3, 4)
+from . import util
+
+
+def in_context(context, f=None):
+    '''Decorator to run a function in a context.
+
+    Context arguments are added as position-only arguments to the signature of
+    the decorated function. Any overlap between parameters of the function and
+    context results in a ``ValueError``.'''
+
+    if f is None:
+        return lambda f: in_context(context, f)
+
+    sig = inspect.signature(f)
+    params = list(sig.parameters.values())
+    cnames = []
+    for param in inspect.signature(context).parameters.values():
+        if param.kind == param.POSITIONAL_OR_KEYWORD:
+            param = param.replace(kind=param.KEYWORD_ONLY)
+            # kind serves to make sure that we can always append parameters to the existing signature
+        elif param.kind != param.KEYWORD_ONLY:
+            raise Exception(f'context parameter {param.name!r} cannot be specified as keyword argument')
+        params.append(param)
+        cnames.append(param.name)
 
+    def in_context(*args, **kwargs):
+        with context(**{name: kwargs.pop(name) for name in cnames if name in kwargs}):
+            f(*args, **kwargs)
+
+    in_context.__signature__ = sig.replace(parameters=params)
+    in_context.__wrapped__ = f
+    in_context.__doc__ = f.__doc__
```

Comment:
Why not use `@functools.wraps(f)` here. See earlier comment.

Reply:
fixed